pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -142,6 +177,174 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+            UnaryExpression.Operator.BITWISE_NOT,
+            UnaryExpression.Operator.PRE_INCREMENT,
+            UnaryExpression.Operator.PRE_DECREMENT,
+            -> ASTPrefixExpression(node)
+            UnaryExpression.Operator.POST_INCREMENT,
+            UnaryExpression.Operator.POST_DECREMENT,
+            -> ASTPostfixExpression(node)
+        }.apply { buildChildren(node, parent = this) }
+
+    /** Builds an [ASTVariableExpression] wrapper for the [FieldExpression] node. */
+    private fun buildFieldExpression(node: FieldExpression): ASTVariableExpression {
+        val (components, receiver, isSafe) = flattenExpression(node)
+        return ASTVariableExpression(components.last()).apply {
+            buildReferenceExpression(components.dropLast(1), receiver, isSafe).also { it.setParent(this) }
+        }
+    }
+
+    /** Builds an [ASTVariableExpression] wrapper for the [VariableExpression] node. */
+    private fun buildVariableExpression(node: VariableExpression): ASTVariableExpression {
+        val (components, receiver, isSafe) = flattenExpression(node)
+        return ASTVariableExpression(components.last()).apply {
+            buildReferenceExpression(components.dropLast(1), receiver, isSafe).also { it.setParent(this) }
+        }
+    }
+
+    /**
+     * Builds an [ASTMethodCallExpression], [ASTThisMethodCallExpression], or
+     * [ASTSuperMethodCallExpression] wrapper for the [CallExpression] node.
+     */
+    private fun buildCallExpression(node: CallExpression) =
+        when (node.id.string.lowercase()) {
+            "this" -> ASTThisMethodCallExpression(node).apply { buildChildren(node, parent = this) }
+            "super" -> ASTSuperMethodCallExpression(node).apply { buildChildren(node, parent = this) }
+            else -> {
+                val (components, receiver, isSafe) = flattenExpression(node)
+                ASTMethodCallExpression(node).apply {
+                    buildReferenceExpression(components.dropLast(1), receiver, isSafe).also {
+                        it.setParent(this)
+                    }
+                    buildChildren(node, parent = this, exclude = { it == node.receiver })
+                }
+            }
+        }
+
+    /**
+     * Attempts to flatten an [Expression] tree containing variable references into a list of
+     * [Identifier]s. Applicable nodes are [FieldExpression], [CallExpression], and
+     * [VariableExpression].
+     *
+     * @return [Triple] of
+     * 1. List of extracted [Identifier]s
+     * 2. The remaining [Expression] that could not be flattened
+     * 3. Whether a safe [access][FieldExpression.isSafe] or [call][CallExpression.isSafe] was found
+     */
+    private fun flattenExpression(
+        node: Expression?,
+        components: List<Identifier> = emptyList()
+    ): Triple<List<Identifier>, Expression?, Boolean> =
+        when (node) {
+            is FieldExpression ->
+                if (node.isSafe) {
```

Comment:
Is there a reasonable way to write a helper function called "flattenIfSafe" that can be called by all `is` branches here?

Reply:
Resolved: N/A after [3fb57fa](https://github.com/eklimo/pmd/pull/8/commits/3fb57fa9d4ec31c55b1481417f53b0e179ec66e2)